// @apollo/client/react/index.compiled.js
export { getApolloContext } from "./context/ApolloContext.js";
export { ApolloProvider } from "./context/ApolloProvider.js";
export { useApolloClient } from "./hooks-compiled/useApolloClient.js";
export { useLazyQuery } from "./hooks-compiled/useLazyQuery.js";
export { useMutation } from "./hooks-compiled/useMutation.js";
export { useQuery } from "./hooks-compiled/useQuery.js";
export { useSubscription } from "./hooks-compiled/useSubscription.js";
export { useReactiveVar } from "./hooks-compiled/useReactiveVar.js";
export { useFragment } from "./hooks-compiled/useFragment.js";
export { useSuspenseQuery } from "./hooks-compiled/useSuspenseQuery.js";
export { useBackgroundQuery } from "./hooks-compiled/useBackgroundQuery.js";
export { useSuspenseFragment } from "./hooks-compiled/useSuspenseFragment.js";
export { useLoadableQuery } from "./hooks-compiled/useLoadableQuery.js";
export { useQueryRefHandlers } from "./hooks-compiled/useQueryRefHandlers.js";
export { useReadQuery } from "./hooks-compiled/useReadQuery.js";
export { skipToken } from "./hooks-compiled/constants.js";
export { createQueryPreloader } from "./query-preloader/createQueryPreloader.js";
export const reactCompilerVersion = "1.0.0";
