// @apollo/client/__cjs/link/index.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApolloLink = exports.execute = exports.concat = exports.split = exports.from = exports.empty = void 0;
var empty_js_1 = require("./core/empty.cjs");
Object.defineProperty(exports, "empty", { enumerable: true, get: function () { return empty_js_1.empty; } });
var from_js_1 = require("./core/from.cjs");
Object.defineProperty(exports, "from", { enumerable: true, get: function () { return from_js_1.from; } });
var split_js_1 = require("./core/split.cjs");
Object.defineProperty(exports, "split", { enumerable: true, get: function () { return split_js_1.split; } });
var concat_js_1 = require("./core/concat.cjs");
Object.defineProperty(exports, "concat", { enumerable: true, get: function () { return concat_js_1.concat; } });
var execute_js_1 = require("./core/execute.cjs");
Object.defineProperty(exports, "execute", { enumerable: true, get: function () { return execute_js_1.execute; } });
var ApolloLink_js_1 = require("./core/ApolloLink.cjs");
Object.defineProperty(exports, "ApolloLink", { enumerable: true, get: function () { return ApolloLink_js_1.ApolloLink; } });
